app/controllers/proposals_controller.rb

```diff
@@ -148,4 +149,28 @@ def require_waitlisted_or_accepted_state
       redirect_to event_url(@event.slug)
     end
   end
+
+  def require_complete_profile
+    return if current_user.complete? && current_user.unconfirmed_email.blank?
+
+    @incomplete_profile_msg = construct_incomplete_profile_msg.html_safe
+  end
+
+  def construct_incomplete_profile_msg
```

Comment:
I'd rename this to be incomplete_profile_msg since we're now using it inline.  No need to tack on construct_.

This method has a serious case of feature envy.  Additionally, it feels like half this method is duplicating AR error message functionality.  Though we do want to add in additionally messaging, we can move this behavior to the model, assuming we can't easily use AR error messages.

Now, the User class is overcomplicated, especially around validations.  For simplicity, I would make a new method on User called profile_errors.  We can have it mimic AR.errors so we can use it in the same way.  The other option, which might be totally fine is to have profile_complete? generate errors if false just like valid?.

For fun, I went ahead and wrote a method to try out it.  I'm not sure whether I would have complete? generate this automatically or if I would have them be separate methods.

```
  def profile_errors
    profile_errors = ActiveModel::Errors.new(self)
    profile_errors.add(:name, :blank, message: "can't be blank") if name.blank?
    if unconfirmed_email.present?
      profile_errors.add(:email, :unconfirmed, message: "#{unconfirmed_email} needs confirmation")
    else
      profile_errors.add(:email, :blank, message: "can't be blank") if email.blank?
    end
    profile_errors
  end
```
It seems like you could then do something like this:

```
  def incomplete_profile_msg
    if profile_errors = current_user.profile_errors
      msg = "Before submitting a proposal your profile needs completing. Please correct the following: "
      msg << profile_errors.full_messages.to_sentence
      msg << ". Visit #{view_context.link_to('My Profile', edit_profile_path)} to update."
      msg.html_safe
    end
  end
```